### PR TITLE
[RLlib] Prototype: Model Trajectory View API, part 0

### PR DESCRIPTION
**Pull request by sven1977**
This PR introduces a suggested Policy/Model Trajectory View API, in which the Policy queries the Model before each call, which time-slices of some trajectory data (obs, actions, rewards, anything the model previously output, etc..) which the Model requires for the upcoming forward pass (inference or training).

This PR does not add any new functionality.

PR aims at eventually being able to retire:
- state and seq_len inputs
- NEXT_OBS field in SampleBatch
- prev_actions/reward
- caching options for previously seen observations in current attention net implementations.
- Atari frame stacking.

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/latest/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested (please justify below)

### CODE REVIEW

File: rllib/models/modelv2.py [hunk 1]
```diff
@@ -223,6 +232,26 @@ def from_batch(self, train_batch, is_training=True):
             i += 1
         return self.__call__(input_dict, states, train_batch.get("seq_lens"))
 
+    def get_view_requirements(self, is_training=False):
```
Comment (ericl) [inline rllib/models/modelv2.py:235]: Can you mark this as experimental?

File: rllib/models/modelv2.py [hunk 2]
```diff
@@ -223,6 +232,26 @@ def from_batch(self, train_batch, is_training=True):
...
+            is_training (bool): Whether the returned requirements are for
+                training or inference (default).
+
+        Returns:
+            Optional[Dict[str, Dict]]: The view requirements as a dict mapping
+                column names e.g. "obs" to config dicts containing supported
+                fields.
+                TODO: (sven) Currently only `timesteps==0` can be setup.
+        """
+        return None
```
Comment (ericl) [inline rllib/models/modelv2.py:253]: Could this return a default config instead of None? Then it doesn't need to be optional type either.

File: rllib/models/modelv2.py [hunk 3]
```diff
@@ -223,6 +232,26 @@ def from_batch(self, train_batch, is_training=True):
             i += 1
         return self.__call__(input_dict, states, train_batch.get("seq_lens"))
 
+    def get_view_requirements(self, is_training=False):
+        """Returns a list of ViewRequirements for this Model (or None).
+
+        A ViewRequirement object tells the caller of this Model, which
```
Comment (ericl) [inline rllib/models/modelv2.py:238]: Shall we try a namedtuple or class instead of raw dicts here? Since this is a python only API, no need to be Json compatible.

File: rllib/policy/policy.py [hunk 4]
```diff
@@ -430,3 +466,36 @@ def map_(a, s):
         return a
 
     return tree.map_structure(map_, action, action_space)
+
+
+def get_trajectory_view(
+    model: ModelV2,
```
Comment (ericl) [inline rllib/policy/policy.py:472]: Move this to another file?

File: rllib/policy/policy.py [hunk 5]
```diff
@@ -189,6 +192,39 @@ def compute_single_action(self,
         return single_action, [s[0] for s in state_out], \
             {k: v[0] for k, v in info.items()}
 
+    @DeveloperAPI
+    def _compute_actions_from_trajectories(
```
Comment (ericl) [inline rllib/policy/policy.py:196]: Can you remove.the underscore, but mark this as experimental for now instead of developerapi?

File: rllib/policy/torch_policy.py [hunk 6]
```diff
@@ -185,6 +185,84 @@ def compute_actions(self,
...
+            **kwargs):
+
+        explore = explore if explore is not None else self.config["explore"]
+        timestep = timestep if timestep is not None else self.global_timestep
+
+        with torch.no_grad():
+            # Create a view and pass that to Model as `input_dict`.
+            input_dict = self._lazy_tensor_dict(
+                get_trajectory_view(self.model, trajectories, is_training=False))
+            # TODO: (sven) support RNNs w/ fast sampling.
```
Comment (ericl) [inline rllib/policy/torch_policy.py:205]: Isn't the only difference from the other compute actions the way we generate the input dict? Does it make sense to reuse the rest of the code after that point?

File: rllib/models/modelv2.py (hunk 1 above)
Comment (sven1977) [reply rllib/models/modelv2.py:235]: +1

File: rllib/policy/torch_policy.py (hunk 6 above)
Comment (sven1977) [reply rllib/policy/torch_policy.py:205]: We can share the code, sure.

File: rllib/policy/policy.py (hunk 5 above)
Comment (sven1977) [reply rllib/policy/policy.py:196]: +1

File: rllib/policy/policy.py (hunk 4 above)
Comment (sven1977) [reply rllib/policy/policy.py:472]: +1

File: rllib/models/modelv2.py (hunk 3 above)
Comment (sven1977) [reply rllib/models/modelv2.py:238]: Used to be a dataclass. I can change it to a `Dict[str, dataclass-object]`, where str is the column name in the `input_dict`.

File: rllib/policy/policy.py (hunk 4 above)
Comment (sven1977) [reply rllib/policy/policy.py:472]: done

File: rllib/models/modelv2.py (hunk 3 above)
Comment (sven1977) [reply rllib/models/modelv2.py:238]: done

File: rllib/models/modelv2.py (hunk 2 above)
Comment (sven1977) [reply rllib/models/modelv2.py:253]: done

Comment (sven1977) [reply rllib/models/modelv2.py:253]: done

File: rllib/models/modelv2.py (hunk 1 above)
Comment (sven1977) [reply rllib/models/modelv2.py:235]: done

Comment (sven1977) [reply rllib/models/modelv2.py]: done

File: rllib/models/modelv2.py (hunk 2 above)
Comment (sven1977) [reply rllib/models/modelv2.py]: done

File: rllib/policy/policy.py (hunk 5 above)
Comment (sven1977) [reply rllib/policy/policy.py]: done

File: rllib/models/modelv2.py [hunk 7]
```diff
@@ -223,6 +241,33 @@ def from_batch(self, train_batch, is_training=True):
...
+
+        Returns:
+            Dict[str, ViewRequirement]: The view requirements as a dict mapping
+                column names e.g. "obs" to config dicts containing supported
+                fields.
+                TODO: (sven) Currently only `timesteps==0` can be setup.
+        """
+        # Default implementation for simple RL model:
+        # Single requirement: Pass current obs as input.
+        return {
```
Comment (ericl) [inline rllib/models/modelv2.py:267]: We'll need to add prev obs and reward but this is fine for now.

File: rllib/policy/policy.py [hunk 8]
```diff
@@ -189,6 +191,39 @@ def compute_single_action(self,
         return single_action, [s[0] for s in state_out], \
             {k: v[0] for k, v in info.items()}
 
+    @ExperimentalAPI
+    def compute_actions_from_trajectories(
```
Comment (ericl) [inline rllib/policy/policy.py:195]: Can we omit experimental? I think a comment is sufficient since thru are temporary and rare.

File: rllib/policy/torch_policy.py (hunk 6 above)
Comment (ericl) [reply rllib/policy/torch_policy.py:207]: Can we do this in this pr?

**Comment by ericl** [comment]
Main comment is can we fix the copy paste in compute_actions()?

**Review by ericl — APPROVED** [review]

File: rllib/models/modelv2.py [hunk 9]
```diff
@@ -66,6 +78,7 @@ def get_initial_state(self):
         """
         return []
 
+    @PublicAPI
     def forward(self, input_dict, state, seq_lens):
```
Comment (ericl) [inline rllib/models/modelv2.py:82]: We have been inconsistent about whether to mark methods as public api or not, if the entire class is PublicAPI (I believe currently usually not). Should we favor one or the other?

I don't think the current documentation for PublicAPI mentions this.

Comment (sven1977) [reply rllib/models/modelv2.py:82]: Yes, we should favor marking the particular methods as well. I think that's more explicit. I would even say we do this in the child classes (and their methods). It's confusing otherwise.

**Comment by richardliaw** [comment]
@sven1977 this PR breaks Jenkins; can you please fix?
